### ISSUE TITLE
Getting Garbage transcription while using our custom model for nodejs streaming

### ISSUE THREAD

**Issue by kjzing** [mozilla/DeepSpeech#3171]
Hi Team,

We have built our custom deepspeech models for version 0.6. We tried using nodejs streaming example for pretrained model for DS 0.6. This works perfectly fine. But as we put our models on the same setup, the transcription start to come as garbage or any word which keep on repeating.
This is happening with Nodejs streaming example as well as Python base example. Is there anything specific have to be done while creating the model for streaming.
Our model works perfectly in case of normal inference. 

Do let me know if any specific information is needed from our side or please point to any specific resource which we could have missed out. We have tried multiple options but of no use.

Thanks in Advance
Kj

**Comment by kjzing** [comment]
Thanks @lissyx 

Excuse me that I didn't put much of the details here. I am elaborating the steps here, it could happen there might be some issue with the examples and compatibility as well.

_How we created **our deepspeech model** and done inference, this is working fine for us._

Data:1500 hrs (8k audio files)

1. The files are stereo format files, created separate channels using command 
sox input.wav output.wav remix 1 (for left) and sox input.wav output.wav remix 2 (for right)

2. Normalize the left and right audio for peak normalized, using :
 
ffmpeg-normalize input_wav_chunk.wav --normalization-type peak --target-level -14 --output norm_wav_chunk.wav

3. Convert normalized left and right wav into chunks (based on start_time, end_time written) using pydub.

audio = AudioSegment.from([audio_cal.wav])
audio_chunk = audio[ int((start_time-0.05)*1000) : int((end_time+0.05)*1000) ] where 0.05 sec(500 ms buffer) buffer is uesd for smooth audio cut
chunk_size may vary between 1.5-15 sec.

4. Create_CSV (Train,test,dev)

5. Create LM using kenlm to make language model of around 300k lines.

command:
../kenlm/build/bin/lmplz --text vocab.txt --arpa words.arpa --o 4 --discount_fallback --prune 0 0 1
../kenlm/build/bin/build_binary -a 255 -q 8 -v trie words.arpa lm.binary
take a pull of 0.6.1 native client extract and then used:
./generate_trie alphabet.txt lm.binary trie 


6. Training Command

python3 DeepSpeech.py \
--alphabet_config_path data/alphabet.txt \
--checkpoint_dir checkpoint_dir \
--train_files data/train.csv \  
--dev_files data/dev.csv \
--test_files data/test.csv \
--epochs 20 \
--use_allow_growth True \
--use_cudnn_rnn True \
--report_count 10 \
--show_progressbar true \
--train_batch_size 16 \
--dev_batch_size 24 \
--test_batch_size 32 \
--learning_rate 0.0001 \
--dropout_rate 0.20 \
--n_hidden 2048 \
--audio_sample_rate 8000 \
--export_dir data/export_dir \
--lm data/lm/lm.binary \
--trie data/lm/trie \
--lm_alpha 0.75 \
--lm_beta 1.85 \

7. **Inference**: then used deepspeech binary for inference.

_This is working fine_
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------

_Using Deepspeech streaming example_

git clone -b r0.6 https://github.com/mozilla/DeepSpeech-examples.git

1. cd mic_vad_streaming
source $HOME/tmp/deepspeech-venv-0.6.1
sudo apt install portaudio19-dev python3-dev
pip install -r requirements.txt

Changes done in mic_vad_streaming.py

BEAM_WIDTH=1024, DEFAULT_SAMPLE_RATE = 8000.

python mic_vad_streaming.py -m ../mic_vad_streaming/deepspeech-0.6.1-models-cd/output_graph.pbmm -l ../mic_vad_streaming/deepspeech-0.6.1-models-cd/lm.binary -t ../mic_vad_streaming/deepspeech-0.6.1-models-cd/trie

_Used our custom model, lm and we are getting garbage output but then the pre trained model, English model which comes with deepspeech output is good._


2. cd web_microphone_websocket
yarn install
yarn start

changes done in server.js:
line 41: vad.processAudio(data, 16000)-------------->vad.processAudio(data, 8000)
line 191: recordedAudioLength += (chunk.length / 2) * (1 / 16000) * 1000; ------------------->recordedAudioLength += (chunk.length / 2) * (1 / 8000) * 1000;
line 19 and line 20:changed model and lm paths as our custom model and lm.

In a separate window run this command
node server.js

_Here also we get garbage outout for our custom model but pretrained model gave good output_

**Comment by lissyx** [comment]
> Thanks @lissyx
> 
> Excuse me that I didn't put much of the details here. I am elaborating the steps here, it could happen there might be some issue with the examples and compatibility as well.
> 
> _How we created **our deepspeech model** and done inference, this is working fine for us._
> 
> Data:1500 hrs (8k audio files)
> 
>     1. The files are stereo format files, created separate channels using command
>        sox input.wav output.wav remix 1 (for left) and sox input.wav output.wav remix 2 (for right)
> 
>     2. Normalize the left and right audio for peak normalized, using :
> 
> 
> ffmpeg-normalize input_wav_chunk.wav --normalization-type peak --target-level -14 --output norm_wav_chunk.wav
> 
>     1. Convert normalized left and right wav into chunks (based on start_time, end_time written) using pydub.
> 
> 
> audio = AudioSegment.from([audio_cal.wav])
> audio_chunk = audio[ int((start_time-0.05)*1000) : int((end_time+0.05)*1000) ] where 0.05 sec(500 ms buffer) buffer is uesd for smooth audio cut
> chunk_size may vary between 1.5-15 sec.
> 
>     1. Create_CSV (Train,test,dev)
> 
>     2. Create LM using kenlm to make language model of around 300k lines.
> 
> 
> command:
> ../kenlm/build/bin/lmplz --text vocab.txt --arpa words.arpa --o 4 --discount_fallback --prune 0 0 1
> ../kenlm/build/bin/build_binary -a 255 -q 8 -v trie words.arpa lm.binary
> take a pull of 0.6.1 native client extract and then used:
> ./generate_trie alphabet.txt lm.binary trie
> 
>     1. Training Command
> 
> 
> python3 DeepSpeech.py
> --alphabet_config_path data/alphabet.txt
> --checkpoint_dir checkpoint_dir
> --train_files data/train.csv \
> --dev_files data/dev.csv
> --test_files data/test.csv
> --epochs 20
> --use_allow_growth True
> --use_cudnn_rnn True
> --report_count 10
> --show_progressbar true
> --train_batch_size 16
> --dev_batch_size 24
> --test_batch_size 32
> --learning_rate 0.0001
> --dropout_rate 0.20
> --n_hidden 2048
> --audio_sample_rate 8000
> --export_dir data/export_dir
> --lm data/lm/lm.binary
> --trie data/lm/trie
> --lm_alpha 0.75
> --lm_beta 1.85 \
> 
>     1. **Inference**: then used deepspeech binary for inference.
> 
> 
> ## _This is working fine_
> 
> _Using Deepspeech streaming example_
> 
> git clone -b r0.6 https://github.com/mozilla/DeepSpeech-examples.git
> 
>     1. cd mic_vad_streaming
>        source $HOME/tmp/deepspeech-venv-0.6.1
>        sudo apt install portaudio19-dev python3-dev
>        pip install -r requirements.txt
> 
> 
> Changes done in mic_vad_streaming.py
> 
> BEAM_WIDTH=1024, DEFAULT_SAMPLE_RATE = 8000.
> 
> python mic_vad_streaming.py -m ../mic_vad_streaming/deepspeech-0.6.1-models-cd/output_graph.pbmm -l ../mic_vad_streaming/deepspeech-0.6.1-models-cd/lm.binary -t ../mic_vad_streaming/deepspeech-0.6.1-models-cd/trie
> 
> _Used our custom model, lm and we are getting garbage output but then the pre trained model, English model which comes with deepspeech output is good._
> 
>     1. cd web_microphone_websocket
>        yarn install
>        yarn start
> 
> 
> changes done in server.js:
> line 41: vad.processAudio(data, 16000)-------------->vad.processAudio(data, 8000)
> line 191: recordedAudioLength += (chunk.length / 2) * (1 / 16000) * 1000; ------------------->recordedAudioLength += (chunk.length / 2) * (1 / 8000) * 1000;
> line 19 and line 20:changed model and lm paths as our custom model and lm.
> 
> In a separate window run this command
> node server.js
> 
> _Here also we get garbage outout for our custom model but pretrained model gave good output_

Again, this is a support request. Use Discourse.

**Comment by lissyx** [comment]
> git clone -b r0.6 https://github.com/mozilla/DeepSpeech-examples.git

Those examples are here to help demonstrate, we can't guarantee they work perfectly ...

**Comment by lissyx** [comment]
> Convert normalized left and right wav into chunks (based on start_time, end_time written) using pydub.

Unclear what you do here. Do you convert to 8k mono or 8k stereo ? All those transformations might alter your model ...

**Comment by kjzing** [comment]
Got it @lissyx, will use Discourse. My idea of putting here is, the pretrained model works well with Streaming API. Our model which is working well in normal inference and gives good output, doesn't work with the same setup of Streaming API, we just change the model location

This is confusing me..

**Comment by lissyx** [comment]
>  doesn't work with the same setup of Streaming API

No, you get poor results with an unsupported example that makes use of the Streaming API in an unsupported usecase on an old version

It would really be helpful if you also trained on 0.7 version, and you test the streaming api with our supported code: C++ client, https://github.com/mozilla/DeepSpeech/blob/v0.7.4/native_client/args.h#L58

**Comment by kjzing** [comment]
So, you mean to say i train my model on DS 0.7 and use DS Streaming Example 0.7, this should work ? Or i have to make certain changes in the streaming example ?

**Comment by lissyx** [comment]
> i train my model on DS 0.7

yes



>  DS Streaming Example 0.7, this should work

no, test against the c++ client using streaming feature, as I linked. that's code we support.
